sycl/doc/extensions/proposed/sycl_ext_intel_fpga_mem.asciidoc

```diff
@@ -0,0 +1,504 @@
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Representation of FPGA memory
+
+[source,c++]
+----
+namespace sycl::ext::intel::experimental {
+template <typename T, typename PropListT = sycl::ext::oneapi::properties<>>
+class fpga_mem {
+  ...
+----
+
+`fpga_mem` is a class template, parameterized by the type of the underlying
+allocation _T_, and a list of properties _PropListT_. The type of the
+allocation _T_ also encodes the size of the allocation.
+
+`fpga_mem` initializes the underlying object `T` with the values passed into its constructor.
+
+Properties may be specified for a `fpga_mem` to provide optimization hints to the compiler. See below for a list of the properties that are allowed.
+
+[NOTE]
+====
+
+`fpga_mem` has similar semantics to a reference wrapper. An fpga_mem templated on a type T can implicitly convert to T, thereby allowing it to be used in most places that an object of type T could be used. The dot operator
+(`operator.`) cannot be overloaded, so a `get()` member is provided to allow a
+reference to be extracted directly when needed.
+====
+
+=== FPGA Memory Interface
+
+The section below and the table following describe the constructors and member functions for `fpga_mem`.
+
+[source,c++]
+----
+namespace sycl::ext::intel::experimental {
+
+  template <typename T, typename PropertyListT = properties<>>
+  class fpga_mem {
+  public:
+    template <typename... S> constexpr fpga_mem(S... args);
+
+    // Access the underlying data
+    operator T &() noexcept;
+    operator const T &() const noexcept;
+
+    T &get() noexcept;
+    const T &get() const noexcept;
+
+    // Enable assignments from underlying type
+    fpga_mem &operator=(const T &) noexcept;
+
+    // Note that there is no need for "fpga_mem" to define member functions
+    // for operators like "++", "[]", "->", comparison, etc. Instead, the type
+    // "T" need only define these operators as non-member functions. Because
+    // there is an implicit conversion from "fpga_mem" to "T&", the operations
+    // can be applied to objects of type "fpga_mem<T>"
+
+    template <typename propertyT> static constexpr bool has_property();
+
+    // The return type is an unspecified internal class used to represent
+    // instances of propertyT
+    template <typename propertyT>
+    static constexpr /*unspecified*/ get_property();
+  };
+
+} // namespace sycl::ext::intel::experimental
+----
+
+[frame="topbot",options="header"]
+|===
+|Functions |Description
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename ... S>
+constexpr fpga_mem(S ... args);
+----
+|
+Constructs a `fpga_mem` object, and implicit storage for `T`.
```

Comment:
As with fpga_datapath, it's worth mentioning that this calls Ts constructor with args....